### DISCUSSION TITLE
Geolocation JS

### DISCUSSION THREAD

**Discussion by adamjehu** [Budibase/budibase#8920] [Help]
Hello!
I'm trying to fill the latitude and longitude coordinates of a mobile device within two fields - without the user being able to manipulate them, like what is possible with adding a map and marker component. 

My approach was to add a button, and defined under its action is _Update State_ and then _Update Field Value_ using the created key. To get the coordinates, I've been trying to use:

`navigator.geolocation.getCurrentPosition(showPosition);
function showPosition(position) {
  return position.coords.latitude;
}`

which is similarly repeated for the longitude field. I'm running the app via HTTPS and a request for location access is presented and granted. Any ideas as to where I'm going wrong with this or perhaps a better approach?

**Comment by melohagan** [comment]
Hey @adamjehu 

Well you would need a return statement, and for the return statement to be the last statement of you binding, e.g.

```js
function showPosition(position) { return position.coords.latitude; }

return navigator.geolocation.getCurrentPosition(showPosition);
```

With that said, the Budibase JavaScript bindings may not have access to the *navigator* object. 

You may need to build a [custom component](https://docs.budibase.com/docs/custom-component) for this one.

**Reply by adamjehu** [reply]
Hello @melohagan ,
Thank you for responding! With the return statement at the end - I still believe Budibase Javascript bindings may have access to the _navigator_ object since I get a browser request for access to locations services. I just can seem to get the requested coordinate to be entered into a field. Initially I was thinking it could be a datatype issues - but the problem appears with both numeric(8,6) and double precision in my PostgreSQL database.

**Comment by adamjehu** [comment]
@melohagan So I tried using _callback_ and _promise_  to get _geolocation_, since I discovered that it runs async:

`function getLocation(callback) {
  if (navigator.geolocation) {
      var lat_coord = navigator.geolocation.getCurrentPosition(function(position){
        var user_position = {};
        user_position = position.coords.latitude; 
        callback(user_position);
      });
  } else {
      alert("0");
  }
}
getLocation(function(lat_coord){
return lat_coord;
});`

and separately:

`var lat;
var promise1 = new Promise(function(resolve, reject) {
    navigator.geolocation.getCurrentPosition(function(pos){
        lat = pos.coords.latitude
        resolve({lat});
    }) 
})

promise1.then(function(value) {
      return value.lat;
});`

Both work well in JSFiddle, but I still cannot get the returned latitude coordinate to be entered into its respective field. Any ideas where I may have gotten it wrong?

**Reply by melohagan** [reply]
Hey @adamjehu 

As per the docs: https://docs.budibase.com/docs/javascript

> JavaScript within Budibase is server-side synchronous only. Async interactions (such as API calls, or await) aren't supported.

Think this means using a **Promise** isn't going to work either.
